waiter/src/waiter/fallback.clj

```diff
@@ -0,0 +1,159 @@
+                      (reset! fallback-state-atom current-state')
+                      current-state')))
+                (catch Exception e
+                  (log/error e "error in fallback-maintainer")
+                  current-state))]
+          (when next-state
+            (recur next-state)))))
+    {:exit-chan exit-chan
+     :query-chan query-chan}))
+
+(defn- lookup-helper
+  "Helper method to lookup if a service exists by keyword in the fallback state."
+  [fallback-state service-id lookup-key]
+  (-> fallback-state lookup-key (contains? service-id)))
+
+(defn service-exists?
+  "Returns true if the requested service-id exists as per the state in the fallback-state."
+  [fallback-state service-id]
+  (lookup-helper fallback-state service-id :available-service-ids))
+
+(defn service-healthy?
+  "Returns true if the requested service-id has a healthy instance as per the state in the fallback-state."
+  [fallback-state service-id]
+  (lookup-helper fallback-state service-id :healthy-service-ids))
+
+(defn descriptor->service-fallback-period-secs
+  "Retrieves the service-fallback-period-secs for the given descriptor."
+  [default-service-fallback-period-secs descriptor]
+  (or (-> descriptor
+          :waiter-headers
+          (get "x-waiter-service-fallback-period-secs"))
+      (-> #{"service-fallback-period-secs"}
+          (sd/token-sequence->merged-data
+            (get-in descriptor [:sources :token->token-data])
+            (get-in descriptor [:sources :token-sequence]))
+          (get "service-fallback-period-secs")
+          (or default-service-fallback-period-secs))))
+
+(defn retrieve-fallback-descriptor
+  "Computes the fallback descriptor with a healthy instance based on the provided descriptor.
+   Fallback descriptors can only be computed for token-based descriptors.
+   The amount of history lookup for fallback descriptor candidates is limited by search-history-length.
+   Also, the fallback descriptor needs to be inside the fallback period to be returned."
+  [descriptor->previous-descriptor default-service-fallback-period-secs search-history-length fallback-state
+   request-time descriptor]
+  (when (-> descriptor :sources :token-sequence count pos?)
+    (let [{{:keys [token->token-data]} :sources} descriptor
+          service-fallback-period-secs (descriptor->service-fallback-period-secs default-service-fallback-period-secs descriptor)]
+      (when (and (pos? service-fallback-period-secs)
+                 (let [most-recently-modified-token (sd/retrieve-most-recently-modified-token token->token-data)
+                       token-last-update-time (get-in token->token-data [most-recently-modified-token "last-update-time"] 0)]
+                   (->> (t/seconds service-fallback-period-secs)
+                        (t/plus (DateTime. token-last-update-time))
+                        (t/before? request-time))))
+        (loop [iteration 1
+               loop-descriptor descriptor]
```

Comment:
Are descriptors really functions of previous descriptors?  Or are they are functions of the request context and the token data at some point in history?

Reply:
Both - the necessary request context and token-data are encapsulated in the descriptor.